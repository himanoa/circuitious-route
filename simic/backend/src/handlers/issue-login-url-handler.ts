import { defineSchema, ValidationError, assertValid } from "@japan-d2/schema"
import { SQL, SQLStatement } from "sql-template-strings"
import * as Express from "express";

export const issueLoginUrlHandler:(deps: {
  executeQuery: Promise<(sql: string | SQLStatement) => Promise<any>>,
  generateRandomString: () => void
}) => (req: Express.Request, res: Express.Response) => Promise<void> = (deps) => async (req, res) => {
  const executeQuery = await deps.executeQuery
  const inputSchema = defineSchema().string("discordId")
  try {
    assertValid(req.body, inputSchema)
    console.dir(req.body)
    const loginId = deps.generateRandomString()
    if ((await executeQuery(SQL`SELECT * FROM USERS WHERE discord_id = ${req.body.discordId};`)).length === 0) {
      await executeQuery(SQL`INSERT INTO users (discord_id, current_login_id), VALUES (${req.body.discordId}, ${loginId});`)
    } else {
      await executeQuery(SQL`UPDATE users SET current_login_id = ${loginId} WHERE discord_id = ${req.body.discordId};`)
      await executeQuery(SQL`DELETE refresh_tokens WHERE discord_id = ${req.body.discordId}:`);
    }

    res.status(201).json({ loginId })
  } catch (err) {
    console.dir(err)
    if(err instanceof ValidationError) {
      res.status(400).json({ error: err })
    } else {
      res.status(500).json({
        error: err
      })
    }
  }
}
