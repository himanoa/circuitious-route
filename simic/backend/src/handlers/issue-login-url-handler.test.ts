import { issueLoginUrlHandler } from "./issue-login-url-handler"
import { SQLStatement } from "sql-template-strings"
import * as Response from "jest-express/lib/response"
import * as Request from "jest-express/lib/request"

describe("issueLoginUrlHandler", () => {
  describe("when exist discordId", () => {
    const executedQuery: string[] = []
    const dummyDeps = {
      executeQuery: Promise.resolve((sql: SQLStatement | string) => {
        if(typeof sql === "string") {
          executedQuery.push(sql)
          return Promise.resolve([1])
        }
        executedQuery.push(sql.sql)
        return Promise.resolve([1])
      }),
      generateRandomString: () => "foobar"
    }
    
    it("should be valid", async () => {
      const request = new Request.Request("https://example.com")
      request.setBody({
        discordId: 123123123
      })
      const response = new Response.Response()
      await issueLoginUrlHandler(dummyDeps)(request as any, response as any)
      expect(executedQuery).toMatchSnapshot()
      expect(response.status).toBeCalledWith(201)
      expect(response.body).toEqual({ "loginId": "foobar"})
    })
  })
  describe("when not exist discordId", () => {
    const executedQuery: string[] = []
    const dummyDeps = {
      executeQuery: Promise.resolve((sql: SQLStatement | string) => {
        if(typeof sql === "string") {
          executedQuery.push(sql)
          return Promise.resolve([])
        }
        executedQuery.push(sql.sql)
        return Promise.resolve([])
      }),
      generateRandomString: () => "foobar"
    }
    
    it("should be valid", async () => {
      const request = new Request.Request("https://example.com")
      request.setBody({
        discordId: 123123123
      })
      const response = new Response.Response()
      await issueLoginUrlHandler(dummyDeps)(request as any, response as any)
      expect(executedQuery).toMatchSnapshot()
      expect(response.status).toBeCalledWith(201)
      expect(response.body).toEqual({ "loginId": "foobar" }) 
    })
  })

  describe("when invalid body", () => {
    const executedQuery: string[] = []
    const dummyDeps = {
      executeQuery: Promise.resolve((sql: SQLStatement | string) => {
        if(typeof sql === "string") {
          executedQuery.push(sql)
          return Promise.resolve([])
        }
        executedQuery.push(sql.sql)
        return Promise.resolve([])
      }),
      generateRandomString: () => "foobar"
    }
    
    it("should be error", async () => {
      const request = new Request.Request("https://example.com")
      request.setBody({
        invalid: 123123123
      })
      const response = new Response.Response()
      await issueLoginUrlHandler(dummyDeps)(request as any, response as any)
      expect(executedQuery).toMatchSnapshot()
      expect(response.status).toBeCalledWith(400)
      expect(response.body).toMatchSnapshot() 
    })
  })

  describe("when db error", () => {
    const executedQuery: string[] = []
    const dummyDeps = {
      executeQuery: Promise.resolve((sql: SQLStatement | string) => {
        throw Error("DB")
      }),
      generateRandomString: () => "foobar"
    }
    
    it("should be error", async () => {
      const request = new Request.Request("https://example.com")
      request.setBody({
        discordId: 123123123
      })
      const response = new Response.Response()
      await issueLoginUrlHandler(dummyDeps)(request as any, response as any)
      expect(executedQuery).toMatchSnapshot()
      expect(response.status).toBeCalledWith(500)
      expect(response.body).toMatchSnapshot() 
    })
  })
})
