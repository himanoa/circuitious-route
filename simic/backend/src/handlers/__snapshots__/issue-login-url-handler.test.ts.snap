// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`issueLoginUrlHandler when db error should be error 1`] = `Array []`;

exports[`issueLoginUrlHandler when db error should be error 2`] = `
Object {
  "error": [Error: DB],
}
`;

exports[`issueLoginUrlHandler when exist discordId should be valid 1`] = `
Array [
  "SELECT * FROM USERS WHERE discord_id = ?",
  "UPDATE users SET current_login_id = ? WHERE discord_id = ?",
  "DELETE refresh_tokens WHERE discord_id = ?",
]
`;

exports[`issueLoginUrlHandler when invalid body should be error 1`] = `Array []`;

exports[`issueLoginUrlHandler when invalid body should be error 2`] = `
Object {
  "error": ValidationError {
    "argument": "discordId",
    "instance": Object {
      "invalid": 123123123,
    },
    "message": "requires property \\"discordId\\"",
    "name": "required",
    "property": "instance",
    "schema": Object {
      "properties": Object {
        "discordId": Object {
          "type": "string",
        },
      },
      "required": Array [
        "discordId",
      ],
      "type": "object",
    },
    "stack": "instance requires property \\"discordId\\"",
  },
}
`;

exports[`issueLoginUrlHandler when not exist discordId should be valid 1`] = `
Array [
  "SELECT * FROM USERS WHERE discord_id = ?",
  "INSERT INTO users (discord_id, current_login_id), VALUES (?, ?)",
]
`;
